src/base/rss/rss_autodownloadrule.cpp

```diff
@@ -41,50 +41,49 @@
 #include <QStringList>
 
 #include "../preferences.h"
-#include "../tristatebool.h"
 #include "../utils/fs.h"
 #include "../utils/string.h"
 #include "rss_feed.h"
 #include "rss_article.h"
 
 namespace
 {
-    TriStateBool jsonValueToTriStateBool(const QJsonValue &jsonVal)
+    boost::optional<bool> jsonValueToOptional(const QJsonValue &jsonVal)
     {
         if (jsonVal.isBool())
-            return TriStateBool(jsonVal.toBool());
+            return jsonVal.toBool();
 
         if (!jsonVal.isNull())
             qDebug() << Q_FUNC_INFO << "Incorrect value" << jsonVal.toVariant();
 
-        return TriStateBool::Undefined;
+        return boost::none;
     }
 
-    QJsonValue triStateBoolToJsonValue(const TriStateBool &triStateBool)
+    QJsonValue optionalToJsonValue(const boost::optional<bool> &opt)
     {
-        switch (static_cast<int>(triStateBool)) {
-        case 0:  return false;
-        case 1:  return true;
-        default: return QJsonValue();
-        }
+        if (!opt)
+            return QJsonValue();
+        return *opt;
     }
 
-    TriStateBool addPausedLegacyToTriStateBool(int val)
+    boost::optional<bool> addPausedLegacyToOptional(int val)
     {
         switch (val) {
-        case 1:  return TriStateBool::True; // always
-        case 2:  return TriStateBool::False; // never
-        default: return TriStateBool::Undefined; // default
+        case 1:  return true; // always
+        case 2:  return false; // never
+        default: return boost::none; // default
         }
     }
 
-    int triStateBoolToAddPausedLegacy(const TriStateBool &triStateBool)
+    int optionalToAddPausedLegacy(const boost::optional<bool> &opt)
     {
-        switch (static_cast<int>(triStateBool)) {
-        case 0:  return 2; // never
-        case 1:  return 1; // always
-        default: return 0; // default
+        if (opt) {
+            if (*opt)
+                return 1; // always
+            else
+                return 2; // never
         }
+        return 0; // use default
```

Comment:
I think here you can safely do:
```c++
if (opt == true)
    return 1; // always
else if (opt ==false)
    return 2; // never
else
    return 0; // use default
```

A reader might scratch his head at first but afterwards he'll realize what we do here. No?

Reply:
I like the @sledgehammer999 variant (if it can be compiled) except the `else`'s after `return`'s.

Reply:
I think @glassez means:
```c++
if (opt == true)
    return 1; // always
else if (opt == false)
    return 2; // never

return 0; // use default
```

Reply:
```c++
if (opt == true)
    return 1; // always
if (opt == false)
    return 2; // never
return 0; // use default
```

Reply:
Done

Reply:
Sorry I'm late to this, just want to point out it might be dangerous to use this form (in this specific case `optional<bool>`), one might reduce it to `if (opt) return 1; if (!opt) return 2;` which is clearly not the same thing. 
To overcome this pitfall, I personally like to see it as a container, using `operator*` to access it, which is also why I didn't think of using `operator==` for comparing with values.

Reply:
```c++
if (opt) return 1;
if (!opt) return 2;
```
In my opinion, it looks "too ugly" to some sane developer want to write it like this.

In fact, this "extreme case" has forced me to abandon the use of any generic classes (boost::optional etc.). My TriStateBool doesn't allow any confusing implicit conversions.
But I don't want, of course, to prevent these changes because of their further spread to other fields. I just wish we have defined some common policy for its use.
>I didn't think of using `operator==` for comparing with values.

But it's convenient, at least for types other than `bool`. Otherwise `boost::optional` becomes not much clear than using ordinary pointers.

Using `if (optional)` to check for value existence is convient but it's confusing in case of `bool`.